deepmd/dpmodel/train/trainer.py

```
@@ -0,0 +1,673 @@
+                step,
+                task_step_result,
+            )
+            valid_results[task.key] = self.evaluate_validation(
+                task,
+                step,
+                task_step_result,
+            )
+        return train_results, valid_results
+
+    def on_train_begin(self, tasks: TrainingTaskCollection) -> None:
+        """Hook called before the first optimizer step."""
```

Ruff (0.15.18)

[warning] `AbstractTrainer.on_train_begin` is an empty method in an abstract base class, but has no abstract decorator

(B027)